torch/cuda/__init__.py

```diff
@@ -646,8 +646,28 @@ def _parse_visible_devices() -> Union[List[int], List[str]]:
 
     if torch.version.hip:
         hip_devices = os.getenv("HIP_VISIBLE_DEVICES")
+        rocr_devices = os.getenv("ROCR_VISIBLE_DEVICES")
+
+        if rocr_devices is not None:
+            # Mostly required for ROCm to make sure ROCr visible devices
+            # is respected, this ensures we do not return a list of devices
+            # that exceeds the total available supplied via ROCR_VISIBLE_DEVICES
+            var = rocr_devices
+
         if hip_devices is not None:
-            var = hip_devices
+            # If ROCr devices have been set, the hip visible devices would
+            # be a subset of those. HIP_VISIBLE_DEVICES can only contain
```

Comment:
If I understand this correctly, I don't believe this assumption to be true.

For example we can set ROCR_VISIBLE_DEVICES="6,7" and HIP_VISIBLE_DEVICES="1,0"

This informs runtime to expose device 6,7 and the hip application has freedom to set these as they wish. 

And as far as I'm aware if we set 

ROCR_VISIBLE_DEVICES="4" HIP_VISIBLE_DEVICES="4" this would cause issues and HIP_VISIBLE_DEVICES wouldn't see a device 4

Reply:
Also HIP_VISIBLE_DEVICES can accept UUID input, so if a user uses ROCR_VISIBLE_DEVICES to expose certain GPUs and then uses HIP_VISIBLE_DEVICES with UUID input then this would also not be a subset.

Reply:
Were there any issues in the original way I proposed to handle this? https://github.com/pytorch/pytorch/pull/140320/commits/9afc7f29a34080843042607d237780e3570c4511

Reply:
We don't really need to parse ROCR_VISIBLE_DEVICES inside PyTorch this is runtime level control of device selection, but what we need to ensure is that the length of ROCR_VISIBLE_DEVICES is respected to avoid the issue in https://github.com/pytorch/pytorch/issues/140318

Reply:
Hi @jataylo, 
Just wanted to leave a small note, documentation here says `HIP_VISIBLE_DEVICES` only accepts indices: https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html#hip-visible-devices
Maybe that documentation is stale? Thanks!

Reply:
Since 6.2 HIP_VISIBLE_DEVICES supports UUID input, we should reach out to update docs.

Reply:
I pushed an update to this logic.  The current logic was wrong; Jack was correct that we don't need to parse or index into ROCR_VISIBLE_DEVICES at all.

Reply:
Thanks! I just want to confirm the behavior when both environment variables are set before correcting the PR. The problem lies in the fact that the function should return a list of indices/strings, so returning an arbitrary list (albeit of the correct length) would work for getting the number of devices, but may fail in cases where the function's output would actually be used.